File: c/test/sensor_tests.c
#include <stdio.h>
#include <stdlib.h>
#include "minunit.h"
#include "../src/parse-utils.h"

#define KGRN  "\x1B[32m"
#define KRED  "\x1B[31m"

int tests_run = 0;

static char* test_1() {
    struct sensor_data sensor_data = parse_sensor_data("00FF000224070000064c");
    float expected_temperature = 25.5;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    int expected_humidity = 36;
    char* error_humidity = (char*) malloc(64);
    snprintf(error_humidity, 64, "Humidity: expected %d but was %d", expected_humidity, sensor_data.humidity);
    mu_assert(error_humidity, sensor_data.humidity == expected_humidity);
    free(error_humidity);

    int expected_battery = 76;
    char* error_battery = (char*) malloc(64);
    snprintf(error_battery, 64, "Battery: expected %d but was %d", expected_battery, sensor_data.battery);
    mu_assert(error_battery, sensor_data.battery == expected_battery);
    free(error_battery);

    return 0;
}

static char* test_2() {
    struct sensor_data sensor_data = parse_sensor_data("0000010224070000064c");

    float expected_temperature = 25.6;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_3() {
    struct sensor_data sensor_data = parse_sensor_data("00FE000224070000064c");

    float expected_temperature = 25.4;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_4() {
    struct sensor_data sensor_data = parse_sensor_data("0052030224070000064c");

    float expected_temperature = 85.0;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_5() {
    struct sensor_data sensor_data = parse_sensor_data("006F000224070000064c");

    float expected_temperature = 11.1;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_6() {
    struct sensor_data sensor_data = parse_sensor_data("000A000224070000064c");

    float expected_temperature = 1.0;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_7() {
    struct sensor_data sensor_data = parse_sensor_data("0001000224070000064c");

    float expected_temperature = 0.1;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_8() {
    struct sensor_data sensor_data = parse_sensor_data("0000000224070000064c");

    float expected_temperature = 0.0;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

     return 0;
}

static char* test_9() {
    struct sensor_data sensor_data = parse_sensor_data("00FFFF0224070000064c");

    float expected_temperature = -0.1;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_10() {
    struct sensor_data sensor_data = parse_sensor_data("00F6FF0224070000064c");

    float expected_temperature = -1.0;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_11() {
    struct sensor_data sensor_data = parse_sensor_data("0091FF0224070000064c");

    float expected_temperature = -11.1;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_12() {
    struct sensor_data sensor_data = parse_sensor_data("003BFF0224070000064c");

    float expected_temperature = -19.7;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* test_13() {
    struct sensor_data sensor_data = parse_sensor_data("0038FF0224070000064c");

    float expected_temperature = -20.0;
    char* error_temperature = (char*) malloc(64);
    snprintf(error_temperature, 64, "Temperature: expected %.1f but was %.1f", expected_temperature, sensor_data.temperature);
    mu_assert(error_temperature, sensor_data.temperature == expected_temperature);
    free(error_temperature);

    return 0;
}

static char* all_tests() {
    mu_run_test(test_1);
    mu_run_test(test_2);
    mu_run_test(test_3);
    mu_run_test(test_4);
    mu_run_test(test_5);
    mu_run_test(test_6);
    mu_run_test(test_7);
    mu_run_test(test_8);
    mu_run_test(test_9);
    mu_run_test(test_10);
    mu_run_test(test_11);
    mu_run_test(test_12);
    mu_run_test(test_13);

    return 0;
}

int main(int argc, char **argv) {

    char *result = all_tests();
    if (result != 0) {
        printf("%s%s\n", KRED, result);
    }
    else {
        printf("%sALL TESTS PASSED\n", KGRN);
    }
    printf("Tests run: %d\n", tests_run);

    return 0;
}

File: python/Reader/test/SensorTests.py
import unittest

from src.utils.ParseUtils import ParseUtils

class SensorTests(unittest.TestCase):

    def test_1(self):
        bytes = "00 FF 00 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(25.5, sensorData.temperature)
        self.assertEquals(36, sensorData.humidity)
        self.assertEquals(76, sensorData.battery)

    def test_2(self):
        bytes = "00 00 01 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(25.6, sensorData.temperature)

    def test_3(self):
        bytes = "00 FE 00 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(25.4, sensorData.temperature)

    def test_4(self):
        bytes = "00 52 03 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(85.0, sensorData.temperature)

    def test_5(self):
        bytes = "00 6F 00 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(11.1, sensorData.temperature)

    def test_6(self):
        bytes = "00 0A 00 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(1.0, sensorData.temperature)

    def test_7(self):
        bytes = "00 01 00 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(0.1, sensorData.temperature)

    def test_8(self):
        bytes = "00 00 00 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(0.0, sensorData.temperature)

    def test_9(self):
        bytes = "00 FF FF 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(-0.1, sensorData.temperature)

    def test_10(self):
        bytes = "00 F6 FF 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(-1.0, sensorData.temperature)

    def test_11(self):
        bytes = "00 91 FF 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(-11.1, sensorData.temperature)

    def test_12(self):
        bytes = "00 3B FF 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(-19.7, sensorData.temperature)

    def test_13(self):
        bytes = "00 38 FF 02 24 07 00 00 06 4c".split(" ")
        sensorData = ParseUtils.parseSensorData(bytes)
        self.assertEquals(-20.0, sensorData.temperature)

if __name__ == '__main__':
    unittest.main()
